[Distributed] Fix generic actors and distributed funcs returning those generics

ktoso opened the pull request:
Depends on https://github.com/apple/swift/pull/41616

The actual fix: https://github.com/apple/swift/pull/41747/commits/5c54ad2f2a8e9bfb97a5bb31f4c274bd14cef88e

~Once we merge https://github.com/apple/swift/pull/41616 I'll rebase this.~ Done

Resolves rdar://84458365

ktoso committed:
[Distributed] Fix generic returning distributed funcs

ktoso commented:
@swift-ci please smoke test

xedin commented on lib/Sema/CodeSynthesisDistributedActor.cpp:
```diff
@@ -523,10 +525,10 @@ static FuncDecl *createDistributedThunkFunction(FuncDecl *func) {
-                  /*parameterNameLoc=*/SourceLoc(), paramName, DC);
+    auto paramDecl = new (C) ParamDecl(SourceLoc(),
+                               SourceLoc(), funcParam->getArgumentName(),
+                               SourceLoc(), funcParam->getParameterName(),
```
This has to use `paramName`, the parameter cannot be anonymous if it's used in the body.

ktoso replied:
Oh, thanks for catching that

ktoso replied:
I messed up the rebase picking the wrong impl here it seems, thanks

ktoso commented:
@swift-ci please smoke test

ktoso commented:
This actually still fails in SIL handling in the init/deinit when the ACTOR is generic. I'll handle this separately.

ktoso commented:
Okey this actually still has missing symbols that I noticed that I added runtime tests... I'll give it a sleep tonight but will try to make progress tomorrow

ktoso commented:
Note to self: this one over to @DougGregor to see if he can figure it out

DougGregor commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test Linux

xedin committed:
[Distributed] Fix ActorSystem requests on protocols that conform to DistributedActor

ktoso committed:
[Distributed] Also cover remote distributed call in test

ktoso commented:
Thank you a lot @xedin!  🙇 Looking good, I cleaned up a bit and added a minor test just to be sure remote call was fine as well (it was).

@swift-ci please smoke test and merge

xedin commented on lib/SIL/IR/SILFunction.cpp:
```diff
@@ -319,6 +319,8 @@ SILType SILFunction::mapTypeIntoContext(SILType type) const {
 SILType GenericEnvironment::mapTypeIntoContext(SILModule &M,
                                                SILType type) const {
+  if (type.hasArchetype())
+    type.dump();
```
You forgot to remove `dump` :)

ktoso replied:
Oh no 🙈

ktoso replied:
Fixed and had to rebase anyway.

ktoso commented:
@swift-ci please smoke test and merge
